util/failuredomains/failure_domains_test.go

```diff
@@ -172,12 +179,696 @@ func TestNewFailureDomainPickMost(t *testing.T) {
+	machineB1New := &clusterv1.Machine{ObjectMeta: metav1.ObjectMeta{Name: "b1-new"}, Spec: clusterv1.MachineSpec{FailureDomain: ptr.To(b)}}
+	machineC1New := &clusterv1.Machine{ObjectMeta: metav1.ObjectMeta{Name: "c1-new"}, Spec: clusterv1.MachineSpec{FailureDomain: ptr.To(c)}}
+
+	testcases := []struct {
+		name             string
+		fds              clusterv1.FailureDomains
+		allMachines      collections.Machines
+		upToDateMachines collections.Machines
+		expected         []string
+	}{
+		// Use case A: 3 failure domains, 3 control plane machines
+
+		// scenario A.1: scale up to 3
+		{
+			name:             "3 failure domains, 0 machine, scale up from 0 to 1",
+			fds:              fds3,
+			allMachines:      collections.FromMachines(),
+			upToDateMachines: collections.FromMachines(),
+			expected:         []string{a, b, c}, // select fd a, b or c because they all have no up-to-date machines
+		},
+		{
+			name:             "3 failure domains, 1 machine, scale up from 1 to 2",
+			fds:              fds3,
+			allMachines:      collections.FromMachines(machineA1),
+			upToDateMachines: collections.FromMachines(machineA1),
+			expected:         []string{b, c}, // select fd b or c because they all have no up-to-date machines
+		},
+		{
+			name:             "3 failure domains, 2 machines, scale up from 2 to 3",
+			fds:              fds3,
+			allMachines:      collections.FromMachines(machineA1, machineB1),
+			upToDateMachines: collections.FromMachines(machineA1, machineB1),
+			expected:         []string{c}, // select fd c because it has no up-to-date machines
+		},
+
+		// scenario A.2: scale up during a rollout
+		{
+			name:             "3 failure domains, 3 outdated machines, scale up to 4",
+			fds:              fds3,
+			allMachines:      collections.FromMachines(machineA1Old, machineB1Old, machineC1Old),
+			upToDateMachines: collections.FromMachines(),
+			expected:         []string{a, b, c}, // select fd a, b or c because they all have no up-to-date machines, 1 machine overall
+		},
+		{
+			name:             "3 failure domains, 2 outdated machines, 1 new machine, scale up to 4",
+			fds:              fds3,
+			allMachines:      collections.FromMachines(machineA1New, machineB1Old, machineC1Old),
+			upToDateMachines: collections.FromMachines(machineA1New),
+			expected:         []string{b, c}, // select fd b or c because they all have no up-to-date machines (less than a)
+		},
+		{
+			name:             "3 failure domains, 1 outdated machine, 2 new machines, scale up to 4",
+			fds:              fds3,
+			allMachines:      collections.FromMachines(machineA1New, machineB1Old, machineC1Old),
+			upToDateMachines: collections.FromMachines(machineA1New, machineB1New),
+			expected:         []string{c}, // select fd c because it has no up-to-date machines (less than b, c)
+		},
+
+		// scenario A.3: scale up after machines has been remediated or forcefully deleted
```

Comment:
How could it happen that we have two old machines in the same FD?

Reply:
If the failure domain was added after the 2 Machines were created

Reply:
We don't have any test coverage for adding removal of FD, because part of the logic that handles this is outside of this package, e.g. https://github.com/kubernetes-sigs/cluster-api/blob/0061ee79d3820dd62e41216b9fe53a28c5490911/controlplane/kubeadm/internal/control_plane.go#L178-L187

In the PR description I have added "I will open a follow up issue to add tests in the upper layer of the call stack" specifically for adding removal of FD (now clarified)

Reply:
Alright!

Reply:
https://github.com/kubernetes-sigs/cluster-api/issues/11602